lib/smack/ConstantBVOps.cpp

```diff
@@ -0,0 +1,107 @@
+//
+// This file is distributed under the MIT License. See LICENSE for details.
+//
+
+// This pass converts bitwise operations with certain constant operands
+// into equivalent integer operations.
+
+#include "smack/ConstantBVOps.h"
+#include "smack/Debug.h"
+#include "smack/Naming.h"
+#include "smack/SmackOptions.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/ValueSymbolTable.h"
+#include "llvm/Support/Regex.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include <string>
+
+namespace smack {
+
+using namespace llvm;
+
+unsigned getOpWidth(const Value *t) {
+  Type *type = t->getType();
+  IntegerType *iType = dyn_cast<IntegerType>(type);
+  return iType->getBitWidth();
+}
+
+bool ConstantBVOps::runOnFunction(Function &f) {
+  if (!(SmackOptions::BitPrecise || SmackOptions::BitPrecisePointers))
+    // Do not run this pass in bitvector mode.
+    return false;
+  std::vector<Instruction *> instsFrom;
+  std::vector<Instruction *> instsTo;
+  if (Naming::isSmackName(f.getName()))
+    return false;
+  for (inst_iterator I = inst_begin(f), E = inst_end(f); I != E; ++I) {
+    if (I->isShift()) {
+      BinaryOperator *bi = dyn_cast<BinaryOperator>(&*I);
+      Value *amt = bi->getOperand(1);
+      if (ConstantInt *ci = dyn_cast<ConstantInt>(amt)) {
+        const APInt &value = ci->getValue();
+        unsigned opcode = bi->getOpcode();
+        Instruction::BinaryOps op;
+        if (opcode == Instruction::AShr || opcode == Instruction::LShr) {
+          // Shifting right by a constant amount is equivalent to dividing by
+          // 2^amt
+          op = Instruction::SDiv;
```

Comment:
I chose SDiv because it seems like these numbers might more often be signed. There's not a good reason I can think of to choose one or the other.

Reply:
The signed or unsigned division is the same when the rhs is a positive number, right?

Reply:
I think so.

Reply:
I think they are the same in integer modeling. One thing about it is if this were done in bv mode, we would need to differentiate between AShr and LShr, but I don't think this makes sense for bv to begin with.

Reply:
I forgot we have the new models for integers. How are divisions handled?

Reply:
> I forgot we have the new models for integers. How are divisions handled?

They have a wrapper around their arguments to convert them into proper unsigned/signed values.